.github/actions/sonar/action.yml

```
@@ -0,0 +1,250 @@
+
+inputs:
+  sonar-project-key:
+    description: 'SonarQube project key'
+    required: true
+  workdir:
+    description: 'Working directory for the action'
+    required: true
+  docker-compose-file:
+    description: 'Path to the Docker Compose file'
+    required: true
+  comment-token:
+    description: 'GitHub token for commenting on the PR'
+    required: true
+
+runs:
+  using: composite
+  steps:
+    # Cache the main branch project for future PR analysis
+    - name: 'Cache project on main branch'
+      if: github.ref == 'refs/heads/main'
+      shell: bash
+      run: |
+        mkdir -p ./${{ inputs.sonar-project-key }}
+        cp -r ${{ inputs.workdir }}/. ./${{ inputs.sonar-project-key }}/
+        rm -rf ${{ inputs.sonar-project-key }}/node_modules \
+            ${{ inputs.sonar-project-key }}/package-lock.json \
+            ${{ inputs.sonar-project-key }}/target/node \
+            ${{ inputs.sonar-project-key }}/target/*.original \
+            ${{ inputs.sonar-project-key }}/target/*.jar
+
+    - name: 'Store cache (main branch)'
+      if: github.ref == 'refs/heads/main'
+      uses: actions/cache/save@v4
+      with:
+        path: ./${{ inputs.sonar-project-key }}
+        key: application-${{ inputs.sonar-project-key }}-${{ github.sha }}
+
+    # Restore the cache for PR analysis
+    - name: 'Restore project cache for PR analysis'
+      if: github.event_name == 'pull_request'
+      id: restore_cache
+      uses: actions/cache/restore@v4
+      with:
+        path: ./${{ inputs.sonar-project-key }}
+        key: application-${{ inputs.sonar-project-key }}-${{ github.event.pull_request.base.sha }}
+        restore-keys: |
+          application-${{ inputs.sonar-project-key }}-
+
+    # Start SonarQube server using Docker Compose
+    - name: 'Start SonarQube server'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: docker compose -f ${{ inputs.docker-compose-file }} up --build -d --wait
+
+    # Create SonarQube project
+    - name: 'Create SonarQube project'
+      shell: bash
+      run: |
+        curl -s -u admin:admin -X POST "http://localhost:9000/api/projects/create?name=${{ inputs.sonar-project-key }}&project=${{ inputs.sonar-project-key }}" || true
+
+    # Run SonarQube analysis on the main branch
+    - name: 'Run SonarQube analysis on main branch'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: >-
+        ./mvnw --batch-mode initialize org.jacoco:jacoco-maven-plugin:prepare-agent sonar:sonar
+        -Dsonar.host.url=http://localhost:9000
+        -Dsonar.projectKey=${{ inputs.sonar-project-key }}
+        -Dsonar.login=admin
+        -Dsonar.password=admin
+        -Dsonar.branch.name=main
+      working-directory: ${{ inputs.sonar-project-key }}
+
```

This may add an empty scan to sonar.
```suggestion

    - name: Add an empty scan to sonar main branch
      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit != 'true'
      shell: bash
      run: >-
        docker run --rm sonarsource/sonar-scanner-cli
        -Dsonar.host.url=http://host.docker.internal:9000
        -Dsonar.projectKey=${{ inputs.sonar-project-key }}
        -Dsonar.login=admin
        -Dsonar.password=admin
        -Dsonar.branch.name=main

```

---

Unfortunately, this approach won’t work as intended because SonarQube relies on Git to identify changes between the base and the head of the branch. It needs to know both the base.sha (from the target branch) and the head.sha (from the current branch) to properly analyze and compare the code.

---

I just pushed similar change, let's see if it works. 
And as far as I know `host.docker.internal` cannot be resolved in github actions.

---

It's not being run because you reverted template change, can we get it back?

---

Sure. Can you revert the commit?

---

Yep, will do

---

I’ve had updated the comment to add git and project source volume if needed.

---

I saw it after pushing the change. I'll change it to your suggestion after verifying that it works.